### PR TITLE
[Fetch Migration] Improvements to subprocess handling

### PR DESCRIPTION
**Pull request by kartg**
### Description

The `migration_monitor.py` module now includes a separate function (`monitor_local`) to monitor a Data Prepper subprocess. 

Previously, the migration monitor only interacted with the Data Prepper process via API calls. This has the limitation that the monitor cannot differentiate between the subprocess intermittently not responding (or not yet started) versus a non-running subprocess. Thus, the API calls simply retry indefinitely if no response is received - which resulted in a indefinite polling loop if the subprocess had crashed or ended (say due to a bad configuration, OOM error, etc.)

The new function adds logic to poll the running subprocess and its return code (via Python's `Popen` object) - this is in addition to the monitoring via API calls. Unit tests for these use-cases have been added/updated.

This PR also includes an additional commit which adds a fix to `index_operations.py` to filter out system indices (so they are excluded from metadata migration). Short-circuit logic for empty source clusters has been added, along with related unit tests.

### Testing
```
$ python -m coverage report --omit "*/tests/*"
Name                           Stmts   Miss  Cover
--------------------------------------------------
endpoint_info.py                   6      0   100%
fetch_orchestrator.py             25      0   100%
index_operations.py               37      0   100%
metadata_migration.py            125      1    99%
metadata_migration_params.py       7      0   100%
metadata_migration_result.py       5      0   100%
migration_monitor.py              96      3    97%
migration_monitor_params.py        5      0   100%
utils.py                          13      0   100%
--------------------------------------------------
TOTAL                            319      4    99%

```

### Check List
- [x] New functionality includes testing
  - [x] All tests pass, including unit test, integration test and doctest
- [x] New functionality has been documented
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: FetchMigration/python/migration_monitor.py [hunk 1]
```diff
@@ -65,32 +85,82 @@ def check_if_complete(doc_count: Optional[int], in_flight: Optional[int], no_par
...
+        try:
+            dp_process.wait(timeout=poll_interval_seconds)
+        except subprocess.TimeoutExpired:
+            pass
+        if dp_process.returncode is None:
+            terminal, internal_metrics = check_and_log_progress(endpoint, internal_metrics, target_doc_count)
+    # Loop terminated
+    if not terminal:
+        logging.error("Migration did not complete, process exited with code: " + str(dp_process.returncode))
+        # TODO - Implement rollback
```
Comment (gregschohn) [inline FetchMigration/python/migration_monitor.py:141]: You should tell the user that there's no rollback and that they should consult documentation on how to recover.

File: FetchMigration/python/fetch_orchestrator.py [hunk 2]
```diff
@@ -3,18 +3,19 @@
...
 
 __DP_EXECUTABLE_SUFFIX = "/bin/data-prepper"
 __PIPELINE_OUTPUT_FILE_SUFFIX = "/pipelines/pipeline.yaml"
 
 
-def run(dp_base_path: str, dp_config_file: str, dp_endpoint: str):
+def run(dp_base_path: str, dp_config_file: str, dp_endpoint: str) -> Optional[int]:
     dp_exec_path = dp_base_path + __DP_EXECUTABLE_SUFFIX
     output_file = dp_base_path + __PIPELINE_OUTPUT_FILE_SUFFIX
     metadata_migration_params = MetadataMigrationParams(dp_config_file, output_file, report=True)
```
Comment (gregschohn) [inline FetchMigration/python/fetch_orchestrator.py:21]: Can your code confirm that DataPrepper is bound to only localhost?  I could see a lot of difficulty happening if a second managing agent came along and started acting on it - or if somebody else DDOS'ed the API (while it was causing a lot of enery on the source cluster).

File: FetchMigration/python/migration_monitor_counters.py [hunk 3]
```diff
@@ -0,0 +1,10 @@
+from dataclasses import dataclass
+
+
+@dataclass
+class MigrationMonitorCounters:
+    prev_no_partitions_count: int = 0
+    prev_success_docs: int = -1
+    idle_success_doc_count: int = 0
+    seq_metric_api_fail: int = 0
+    seq_metric_value_fail: int = 0
```
Comment (gregschohn) [inline FetchMigration/python/migration_monitor_counters.py:10]: docs here would be great, especially why we care about each of them, what they mean.

**Review by gregschohn — COMMENTED** [review]
I've added comments to the code as best as I could follow it.  
Some documentation in the main files about what you're doing, what the dependencies are doing, and what the responsibilities are for everything would help make a review from me more valuable.  I don't know if are issues with the current patch because I can't fully assimilate the starting case yet.

File: FetchMigration/python/metadata_migration.py [hunk 4]
```diff
@@ -163,8 +163,12 @@ def run(args: MetadataMigrationParams) -> MetadataMigrationResult:
     # We expect the Data Prepper pipeline to only have a single top-level value
     pipeline_config = next(iter(dp_config.values()))
     validate_pipeline_config(pipeline_config)
+    result = MetadataMigrationResult()
```
[the remark this answers is not in the document]
Comment (kartg) [reply FetchMigration/python/metadata_migration.py:166]: Good spot, thanks!

File: FetchMigration/python/migration_monitor.py [hunk 5]
```diff
@@ -9,6 +11,7 @@
 from prometheus_client.parser import text_string_to_metric_families
 
 from endpoint_info import EndpointInfo
+from migration_monitor_counters import MigrationMonitorCounters
 from migration_monitor_params import MigrationMonitorParams
 
 __PROMETHEUS_METRICS_ENDPOINT = "/metrics/prometheus"
```
[the remark this answers is not in the document]
Comment (kartg) [reply FetchMigration/python/migration_monitor.py]: Added a comment

File: FetchMigration/python/migration_monitor.py (hunk 1 above)
Comment (kartg) [reply FetchMigration/python/migration_monitor.py:133]: Added error logging

File: FetchMigration/python/migration_monitor_counters.py (hunk 3 above)
Comment (kartg) [reply FetchMigration/python/migration_monitor_counters.py]: Sorry, this was bleed from a follow-up PR. This file has now been removed

**Review by sumobrian — APPROVED** [review]
lgtm

**Comment by kartg** [comment]
CodeCov failures are looking at Java code, which this PR doesn't touch. I'll go ahead and merge this.

**Pull request merged**